File: Taggd/TagStore.swift
//
//  TagStore.swift
//  Taggd
//
//  Persistent, ordered library of the user's tags. Shared by the main screen's
//  tag picker and the Settings tag manager. Backed by UserDefaults (JSON).
//

import SwiftUI
import Observation

@Observable
@MainActor
final class TagStore {
    private static let storageKey = "tagLibrary"

    private(set) var tags: [Tag]

    init() {
        if let data = UserDefaults.standard.data(forKey: Self.storageKey),
           let decoded = try? JSONDecoder().decode([Tag].self, from: data) {
            tags = decoded
        } else {
            tags = Tag.presets
        }
    }

    /// Adds a tag (or returns the existing one for a duplicate name). Returns nil for empty input.
    @discardableResult
    func add(_ name: String) -> Tag? {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        if let existing = tags.first(where: { $0.name.caseInsensitiveCompare(trimmed) == .orderedSame }) {
            return existing
        }
        let tag = Tag(name: trimmed)
        tags.append(tag)
        save()
        return tag
    }

    func rename(_ id: Tag.ID, to name: String) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let index = tags.firstIndex(where: { $0.id == id }) else { return }
        guard tags[index].name != trimmed else { return }
        tags[index].name = trimmed
        save()
    }

    func remove(at offsets: IndexSet) {
        tags.remove(atOffsets: offsets)
        save()
    }

    func move(from source: IndexSet, to destination: Int) {
        tags.move(fromOffsets: source, toOffset: destination)
        save()
    }

    private func save() {
        if let data = try? JSONEncoder().encode(tags) {
            UserDefaults.standard.set(data, forKey: Self.storageKey)
        }
    }
}

File: TaggdWin/Services/TagStore.cs
using System;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using System.Text.Json;
using TaggdWin.Models;

namespace TaggdWin.Services
{
    /// <summary>
    /// Persistent, ordered library of the user's tags, shared by the tray popup's
    /// picker and the settings tag manager. Backed by JSON in %APPDATA%\Tagged.
    /// Mirrors the Swift <c>TagStore</c>.
    /// </summary>
    public sealed class TagStore
    {
        public static TagStore Shared { get; } = new TagStore();

        private static readonly string FilePath = Path.Combine(SettingsStore.AppDataDir, "tags.json");

        public ObservableCollection<Tag> Tags { get; }

        private TagStore()
        {
            Tags = Load();
        }

        private static ObservableCollection<Tag> Load()
        {
            try
            {
                if (File.Exists(FilePath))
                {
                    var decoded = JsonSerializer.Deserialize<Tag[]>(File.ReadAllText(FilePath));
                    if (decoded != null)
                        return new ObservableCollection<Tag>(decoded.Where(t => !string.IsNullOrWhiteSpace(t.Name)));
                }
            }
            catch { /* fall back to empty */ }
            return new ObservableCollection<Tag>();
        }

        public void Save()
        {
            try
            {
                Directory.CreateDirectory(SettingsStore.AppDataDir);
                File.WriteAllText(FilePath,
                    JsonSerializer.Serialize(Tags.ToArray(), new JsonSerializerOptions { WriteIndented = true }));
            }
            catch { /* best effort */ }
        }

        /// <summary>Adds a tag (or returns the existing one for a duplicate name). Null for empty input.</summary>
        public Tag? Add(string name)
        {
            var trimmed = (name ?? "").Trim();
            if (trimmed.Length == 0) return null;
            var existing = Tags.FirstOrDefault(t => string.Equals(t.Name, trimmed, StringComparison.OrdinalIgnoreCase));
            if (existing != null) return existing;
            var tag = new Tag(trimmed);
            Tags.Add(tag);
            Save();
            return tag;
        }

        public void Rename(Guid id, string name)
        {
            var trimmed = (name ?? "").Trim();
            if (trimmed.Length == 0) return;
            var tag = Tags.FirstOrDefault(t => t.Id == id);
            if (tag == null || tag.Name == trimmed) return;
            tag.Name = trimmed;
            Save();
        }

        public void Remove(Tag tag)
        {
            if (Tags.Remove(tag)) Save();
        }

        public void Move(int from, int to)
        {
            if (from < 0 || from >= Tags.Count || to < 0 || to >= Tags.Count || from == to) return;
            Tags.Move(from, to);
            Save();
        }
    }
}
